File: backend/services/utils/metrics_extractor.js
// Metriken aus Python-Ausgabe extrahieren
export function extractMetricsFromOutput(output, modelType) {
    const metrics = {};
    const foundMetrics = new Set(); // Track bereits gefundene Metriken

    // Erweiterte Metriken-Extraktionsmuster mit Priorität für ausgeschriebene Namen
    const metricPatterns = [
        {
            primaryName: 'mean_absolute_error',
            aliases: ['mae'],
            regexes: [
                /(?:MAE):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'mean_squared_error',
            aliases: ['mse'],
            regexes: [
                /(?:MSE):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'root_mean_squared_error',
            aliases: ['rmse'],
            regexes: [
                /(?:RMSE):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'r_squared',
            aliases: ['r2'],
            regexes: [
                /(?:R2):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'accuracy',
            aliases: [],
            regexes: [
                /(?:Accuracy):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'precision',
            aliases: [],
            regexes: [
                /(?:Precision)(?:\s*\(Precision\))?:\s*([\d.]+)/i,
                /(?:Precision):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'recall',
            aliases: [],
            regexes: [
                /(?:Recall)(?:\s*\(Recall\))?:\s*([\d.]+)/i,
                /(?:Recall):\s*([\d.]+)/i
            ]
        },
        {
            primaryName: 'f1_score',
            aliases: ['f1'],
            regexes: [
                /(?:F1 Score|F1)(?:\s*\(F1\))?:\s*([\d.]+)/i,
                /(?:F1 Score|F1):\s*([\d.]+)/i
            ]
        }
    ];

    // Extrahiere Metriken mit Priorität für ausgeschriebene Namen
    metricPatterns.forEach(metricGroup => {
        // Prüfe ob diese Metrik bereits gefunden wurde
        const allNames = [metricGroup.primaryName, ...metricGroup.aliases];
        const alreadyFound = allNames.some(name => foundMetrics.has(name));

        if (alreadyFound) {
            return; // Überspringe diese Metrik-Gruppe
        }

        for (const regex of metricGroup.regexes) {
            const match = output.match(regex);
            if (match) {
                const value = parseFloat(match[1]);

                // Validiere den Wert
                if (isNaN(value)) {
                    continue; // Überspringe ungültige Werte
                }

                // Speichere nur den primären Namen (ausgeschrieben)
                metrics[metricGroup.primaryName] = value;

                // Markiere alle Aliase als gefunden
                allNames.forEach(name => foundMetrics.add(name));

                break; // Stoppe nach dem ersten erfolgreichen Match
            }
        }
    });

    return metrics;
}

File: services/unified-service/core/execution/metrics_extractor.py
"""
Metriken-Extraktion aus Python-Ausgabe
"""

import re
from typing import Dict, Any, Optional

def extract_metrics_from_output(output: str, model_type: str) -> Dict[str, Any]:
    """Extrahiert Metriken aus Python-Ausgabe"""
    metrics = {}
    found_metrics = set()
    
    # Metriken-Patterns
    metric_patterns = [
        {
            'primary_name': 'mean_absolute_error',
            'aliases': ['mae'],
            'regexes': [r'(?:MAE):\s*([\d.]+)']
        },
        {
            'primary_name': 'mean_squared_error',
            'aliases': ['mse'],
            'regexes': [r'(?:MSE):\s*([\d.]+)']
        },
        {
            'primary_name': 'root_mean_squared_error',
            'aliases': ['rmse'],
            'regexes': [r'(?:RMSE):\s*([\d.]+)']
        },
        {
            'primary_name': 'r_squared',
            'aliases': ['r2'],
            'regexes': [r'(?:R2|R²|R-squared):\s*([\d.]+)']
        },
        {
            'primary_name': 'accuracy',
            'aliases': [],
            'regexes': [r'(?:Accuracy):\s*([\d.]+)']
        },
        {
            'primary_name': 'precision',
            'aliases': [],
            'regexes': [r'(?:Precision):\s*([\d.]+)']
        },
        {
            'primary_name': 'recall',
            'aliases': [],
            'regexes': [r'(?:Recall):\s*([\d.]+)']
        },
        {
            'primary_name': 'f1_score',
            'aliases': ['f1'],
            'regexes': [r'(?:F1|F1-Score|F1 Score):\s*([\d.]+)']
        }
    ]
    
    # Durchsuche Output nach Metriken
    for pattern in metric_patterns:
        if pattern['primary_name'] in found_metrics:
            continue
        
        for regex_str in pattern['regexes']:
            match = re.search(regex_str, output, re.IGNORECASE)
            if match:
                try:
                    value = float(match.group(1))
                    metrics[pattern['primary_name']] = value
                    found_metrics.add(pattern['primary_name'])
                    print(f'[Metrics Extractor] Gefundene Metrik: {pattern["primary_name"]} = {value}')
                    break
                except (ValueError, IndexError) as e:
                    print(f'[Metrics Extractor] Fehler beim Parsen von {pattern["primary_name"]}: {e}')
                    continue
    
    if not metrics:
        print(f'[Metrics Extractor] WARNUNG: Keine Metriken gefunden im Output (Länge: {len(output)} Zeichen)')
        print(f'[Metrics Extractor] Output-Ausschnitt (letzte 500 Zeichen): {output[-500:]}')
    
    return metrics

